syntax = "proto3";

package tasks;

// IMPORTANTE: O caminho abaixo deve corresponder ao nome da sua pasta principal
// e à estrutura do seu projeto.
option go_package = "lista_de_tarefas_megazord_com_rest_soap_api_mom/go_server/pb";

// Mensagem que representa uma Tarefa
message Task {
  int32 id = 1;
  string title = 2;
  string description = 3;
  string status = 4;
  string created_by = 5;
}

// Mensagens de Requisição e Resposta para os métodos

// Requisição para criar uma nova tarefa
message CreateTaskRequest {
  string title = 1;
  string description = 2;
  string created_by = 3;
}

// Resposta após criar uma tarefa
message CreateTaskResponse {
  Task task = 1;
  string message = 2;
}

// Requisição para listar tarefas
message ListTasksRequest {
  // Nenhum campo para listar todas as tarefas
}

// Resposta para listar tarefas
message ListTasksResponse {
  repeated Task tasks = 1;
  string message = 2;
}

// Requisição para atualizar uma tarefa
message UpdateTaskRequest {
  int32 id = 1;
  string title = 2;
  string description = 3;
  string status = 4;
}

// Resposta após atualizar uma tarefa
message UpdateTaskResponse {
  Task task = 1;
  string message = 2;
}

// Requisição para deletar uma tarefa
message DeleteTaskRequest {
  int32 id = 1;
}

// Resposta após deletar uma tarefa
message DeleteTaskResponse {
  bool success = 1;
  string message = 2;
}

// Requisição para obter uma única tarefa por ID
message GetTaskRequest {
  int32 id = 1;
}

// Resposta para obter uma única tarefa
message GetTaskResponse {
  Task task = 1;
  string message = 2;
}

// Requisição para enviar tarefas por e-mail
message SendTasksByEmailRequest {
  string recipient_email = 1;
}

// Resposta para o envio de e-mail
message SendTasksByEmailResponse {
  bool success = 1;
  string message = 2;
}

// Definição do Serviço de Gerenciamento de Tarefas
service TaskService {
  rpc CreateTask (CreateTaskRequest) returns (CreateTaskResponse);
  rpc ListTasks (ListTasksRequest) returns (ListTasksResponse);
  rpc UpdateTask (UpdateTaskRequest) returns (UpdateTaskResponse);
  rpc DeleteTask (DeleteTaskRequest) returns (DeleteTaskResponse);
  rpc GetTask (GetTaskRequest) returns (GetTaskResponse);
  rpc SendTasksByEmail (SendTasksByEmailRequest) returns (SendTasksByEmailResponse);
}